onnxruntime/core/providers/nuphar/nuphar_execution_provider.cc

```diff
@@ -0,0 +1,384 @@
+      else
+        ORT_ENFORCE(false, "Target except avx2/avx512 are not supported!");
+    } else {
+      const auto& cpu_id_info = CPUIDInfo::GetCPUIDInfo();
+      if (cpu_id_info.HasAVX512f())
+        codegen_target_ = CodeGenTarget_AVX512();
+      else if (cpu_id_info.HasAVX2())
+        codegen_target_ = CodeGenTarget_AVX2();
+      else
+        codegen_target_ = std::make_unique<CodeGenTargetX86>(target_str, 128, 1);  // TODO: use real values
+    }
+  } else {
+    codegen_target_ = std::make_unique<CodeGenTarget>(target_str);
+  }
+
+  CreateTVMTarget();
+
+  tvm_host_target_ = tvm::Target();
+  tvm_ctx_.device_type = static_cast<DLDeviceType>(tvm_target_->device_type);
+  tvm_ctx_.device_id = info.device_id;
+
+  graph_stats_ = std::make_unique<SubgraphPartitionStats>();
+
+  whole_graph_shape_infer_ = std::make_shared<ShapeExprContext>();
+
+  DeviceAllocatorRegistrationInfo allocator_info(
+      {OrtMemTypeDefault,
+       [this](int /*id*/) { return std::make_unique<NupharAllocator>(this->tvm_ctx_); },
+       std::numeric_limits<size_t>::max()});
+
+  InsertAllocator(CreateAllocator(allocator_info, tvm_ctx_.device_id));
+
+  // TODO add multi-target support
+  tvm_codegen_manager_ = std::make_unique<TVMCodeGenManager>();
+
+  // Create codegen handle for one target for now
+  codegen_handles_.clear();
+  auto handle = std::make_unique<NupharCodeGenHandle>();
+  tvm_codegen_manager_->Initialization();
+  tvm_codegen_manager_->SetCodeGenHandle(handle.get());
+  handle->allocator = GetAllocator(0, OrtMemTypeDefault);
+  handle->codegen_target = codegen_target_.get();
+  handle->domain_version_lookup_func =
+      [this](const std::string& domain) {
+        return GetDomainVersion(domain);
+      };
+
+  handle->shape_inference = whole_graph_shape_infer_;
+
+  // TODO: remove
+  handle->enable_per_node_parallelized = info.enable_per_node_parallel;
+  // TODO: remove
+  handle->allow_unaligned_buffers = info.allow_unaligned_buffers;  // TODO remove this
+
+  codegen_handles_.push_back(std::move(handle));
+
+  // Runtime Handle
+  runtime_handle_ = std::make_unique<nuphar::NupharRuntimeHandle>(tvm_ctx_);
+  runtime_handle_->allocator = GetAllocator(0, OrtMemTypeDefault);
+  runtime_handle_->allow_unaligned_buffers = info.allow_unaligned_buffers;
+  runtime_handle_->enable_model_parallelism = false;
+}
+
+Status NupharExecutionProvider::CopyTensor(const Tensor& src, Tensor& dst) const {
+  if (!(strcmp(src.Location().name, TVM_STACKVM) == 0 && strcmp(dst.Location().name, TVM_STACKVM) == 0) &&
+      !(strcmp(src.Location().name, TVM_STACKVM) == 0 && strcmp(dst.Location().name, CPU) == 0) &&
+      !(strcmp(src.Location().name, CPU) == 0 && strcmp(dst.Location().name, TVM_STACKVM) == 0))
+    ORT_NOT_IMPLEMENTED("copy to ", dst.Location().name, " from ", src.Location().name, " is not implemented");
+
+  size_t bytes = src.DataType()->Size() * src.Shape().Size();
+  const void* src_data = src.DataRaw();
```

Comment:
src.SizeInBytes()

Reply:
This function is no longer needed after taking Ke Zhang's change

---
In reply to: [313185597](https://github.com/microsoft/onnxruntime/pull/881#discussion_r313185597) [](ancestors = 313185597)